src/bookshelf/src/bksw_sec.erl

```diff
@@ -20,125 +21,159 @@
+        end,
+    auth_finish(Auth, Context, ComparisonURL, IncomingSig, CalculatedSig).
+
+% @doc split authorization header into component parts
+% https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+-spec parse_authorization(string()) -> {ok, [string()]} | {error, parse_authorization}.
+parse_authorization(Authorization) ->
+    case string:split(Authorization, " ", all) of
+        ["AWS4-HMAC-SHA256", "Credential="++Cred, "SignedHeaders="++SigHead, "Signature="++Signature] ->
+            {ok, [string:trim(Cred, trailing, ","), string:trim(SigHead, trailing, ","), Signature]};
+        _ ->
+            {error, parse_authorization}
     end.
 
-do_signed_url_authorization(RequestId, Req0, #context{reqid = ReqId} = Context) ->
-    AWSAccessKeyId = wrq:get_qs_value("AWSAccessKeyId", Req0),
-    Expires = wrq:get_qs_value("Expires", Req0),
-    IncomingSignature = wrq:get_qs_value("Signature", Req0),
-    RawMethod = wrq:method(Req0),
-    Method = string:to_lower(erlang:atom_to_list(RawMethod)),
-    Headers = mochiweb_headers:to_list(wrq:req_headers(Req0)),
-    Path  = wrq:path(Req0),
-    AccessKey = bksw_conf:access_key_id(Context),
-    SecretKey = bksw_conf:secret_access_key(Context),
-    ExpireDiff = expire_diff(Expires),
-    case make_signed_url_authorization(SecretKey,
-                                       Method,
-                                       Path,
-                                       Expires,
-                                       Headers) of
-        {StringToSign, Signature} ->
-            case ExpireDiff =< 0 of
+encode_access_denied_error_response(RequestId, Req0, Context) ->
+    Req1 = bksw_req:with_amz_id_2(Req0),
+    Body = bksw_xml:access_denied_error(RequestId),
+    Req2 = wrq:set_resp_body(Body, Req1),
+    {{halt, 403}, Req2, Context}.
+
+%%===================================================================
+% local functions, helpers, etc.
+%%===================================================================
+
+% common setup, init
+-spec auth_init(any(), tuple(), string()) -> map().
+auth_init(Req0, Context, SignedHeaders) ->
+    AccessKey            =  bksw_conf:access_key_id(Context),
+    {RequestId, Req1}    =  bksw_req:with_amz_request_id(Req0),
+    Config               =  mini_s3:new(AccessKey, bksw_conf:secret_access_key(Context), host(Req1)),
+    #{accesskey          => AccessKey,
+      config             => Config,
+      alt_signed_headers => [case {K, V} of {"host", _} -> {"host", get_host_toggleport(host(Req1), Config)}; _ -> {K, V} end || {K, V} <- SignedHeaders],
+      method             => list_to_atom(string:to_lower(erlang:atom_to_list(wrq:method(Req1)))),
+      path               => wrq:path(Req1),
+      req                => Req1,
+      reqid              => RequestId}.
+
+% TODO: spec
+auth_finish(Auth, #context{
```

Comment:
might refactor this a bit when the constant time comparison work is done

Reply:
`auth_finish` refactored (removed).